src/utils/bounded_priority_queue.rs

```diff
@@ -0,0 +1,119 @@
+pub(crate) struct BoundedMinPriorityQueue<T: Ord> {
+    max_size: usize,
+    underlying: BinaryHeap<Reverse<T>>,
+}
+
+impl<T: Data + Ord> BoundedMinPriorityQueue<T> {
+    pub fn new(max_size: usize) -> BoundedMinPriorityQueue<T> {
+        BoundedMinPriorityQueue {
+            max_size: max_size,
+            underlying: BinaryHeap::with_capacity(max_size),
+        }
+    }
+
+    pub fn get_size(&self) -> usize {
+        self.underlying.len()
+    }
+
+    pub fn into_vec_sorted(&self) -> Vec<T> {
+        let mut res = self
+            .underlying
+            .clone()
+            .into_iter_sorted()
+            .map(|Reverse(i)| i)
+            .collect::<Vec<_>>();
+        res.reverse();
+        res
+    }
+
+    pub fn merge(&mut self, other: BoundedMinPriorityQueue<T>) -> &Self {
+        other
+            .underlying
+            .into_iter()
+            .for_each(|Reverse(elem)| self.append(elem));
+        self
+    }
+
+    pub fn append(&mut self, elem: T) {
+        if self.underlying.len() < self.max_size {
+            self.underlying.push(Reverse(elem));
+        } else {
+            self.maybe_replace_lowest(elem);
+        }
+    }
+
+    pub(self) fn maybe_replace_lowest(&mut self, elem: T)
+    where
+        T: Data + Ord,
+    {
+        if let Some(Reverse(head)) = self.underlying.peek() {
+            if elem.gt(head) {
+                self.underlying.pop();
+                self.underlying.push(Reverse(elem));
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct BoundedMaxPriorityQueue<T: Ord> {
+    max_size: usize,
+    underlying: BinaryHeap<T>,
+}
+
+impl<T: Data + Ord> BoundedMaxPriorityQueue<T> {
+    pub fn new(max_size: usize) -> BoundedMaxPriorityQueue<T> {
+        BoundedMaxPriorityQueue {
+            max_size: max_size,
+            underlying: BinaryHeap::with_capacity(max_size),
+        }
+    }
+
+    pub fn get_size(&self) -> usize {
+        self.underlying.len()
+    }
+
+    pub fn into_vec_sorted(&self) -> Vec<T> {
+        let mut res = self
+            .underlying
+            .clone()
+            .into_iter_sorted()
+            .collect::<Vec<_>>();
+        res.reverse();
+        res
+    }
+
+    pub fn merge(&mut self, other: BoundedMaxPriorityQueue<T>) -> &Self {
+        other
+            .underlying
+            .into_iter()
+            .for_each(|elem| self.append(elem));
+        self
+    }
+
+    pub fn append(&mut self, elem: T) {
+        if self.underlying.len() < self.max_size {
+            self.underlying.push(elem);
+        } else {
+            self.maybe_replace_lowest(elem);
+        }
+    }
+
+    pub(self) fn maybe_replace_lowest(&mut self, elem: T)
+    where
+        T: Data + Ord,
+    {
+        if let Some(head) = self.underlying.peek() {
+            if elem.lt(head) {
+                self.underlying.pop();
+                self.underlying.push(elem);
+            }
+        }
+    }
+}
```

Comment:
Repetitive `BoundedMinPriorityQueue` and `BoundedMinPriorityQueue`, suggestions are welcome to refactoring them to remove the duplication.

Reply:
As far as I can see the only difference is what is done inside the conversion to a vec, right? Let me think about it, but instead what we can do is return an iterator of a `BoundedPriorityQueue` and do the sorting in the call method (top or take_ordered) and the collect the vec.

Reply:
It is a bit more complicate. 
For `take_ordered()` it uses max-map contained in `BoundedMaxPriorityQueue`. And when it replace the lowest, it use `lt()` to compare the element with **max** value popped out from the BinaryHeap.
However in `top()` it requires a min-heap contained in `BoundedMinPriorityQueue`, and it uses `gt` to compare the element with **min** value popped out. In addition, to build a min-heap with `std::collections::BinaryMap` which by default is a max-heap. It has to use `std::cmp::Reverse<T>` as type paramter like [this](https://github.com/rajasekarv/native_spark/pull/114/files#diff-73d82b076910c5f8f7926af6364192f1R9)

Reply:
Ok let me take a better look at it on how we could express this, but on the top of my head maybe what we want is to have a trait `SortGt` and `SortLt` and implement both for `BoundedPriorityQueue` then you don't call 'into_vec_sorted' but the different methods of each trait in `top` and `take_ordered`.

Reply:
you can simply reverse the order of the Self::Item before sending it to take_ordered. Like this. 
```
Ok(self.map(Fn!(|x| Reverse(x))).take_ordered(num)?.into_iter().map(|x| x.0).collect())
```

The intermediate collection is not that costly as this operation is not supposed to be used for higher numbers. However, if we still care about performance, then just return an iterator rather than Vector.

Reply:
This should remove the duplication if I am not wrong

Reply:
@ugoa If we want to avoid the additional `map` to iterate over all the previous rdd for the solution suggested by raja, one simple fix could be to use a newtype enum and implement PartialOrd for it, like:
```
#[derive(Clone, Debug, Serialize, Deserialize)]
pub(crate) struct BoundedMinPriorityQueue<T: PartialOrd> where T: Data + PartialOrd {
    max_size: usize,
    underlying: BinaryHeap<ReversedOrNot<T>>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
enum ReversedOrNot<T> 
 where T: Data + PartialOrd
{
    Reverse(T),
    NoReverse(T),
}


impl<T: PartialOrd> PartialOrd for ReversedOrNot<T> {
...
}
```

then on `maybe_replace_lowest` you can match on both variants and use lt/gt depending on the variant.

Reply:
@iduartgomez  We don't need need to worry about map iter Replace wrap step as it compiles down to essentially one block only at the end along with bounded priority queue step. Enum option is also basically duplication only. And it also involves the same exact wrapping of T in Reverse(T) as a map function. Why have duplicate functions when only one will do?

Reply:
The map is an extra steep that iterates over all the rdd, I would avoid it, this is what I mean and it works:
https://gist.github.com/56f61a29fdfa1687d5c30257364c4e67

Reply:
then on the fucntion you just do this:
```
fn take_ordered(&self, num: usize, reverse: bool) -> Result<Vec<Self::Item>>
    where
        Self: Sized,
        Self::Item: Data + Ord,
    {
        if num == 0 {
            Ok(vec![])
        } else {
            let first_k_func = Fn!(move |partition: Box<dyn Iterator<Item = Self::Item>>|
                -> Box<dyn Iterator<Item = BoundedPriorityQueue<Self::Item>>>  {
                    let mut queue = BoundedPriorityQueue::new(num);
                    partition.for_each(|item: Self::Item| queue.append_non_reversed(item));
                    Box::new(std::iter::once(queue))
            });

            let queue = self
                .map_partitions(first_k_func)
                .reduce(Fn!(
                    move |queue1: BoundedPriorityQueue<Self::Item>,
                          queue2: BoundedPriorityQueue<Self::Item>|
                          -> BoundedPriorityQueue<Self::Item> {
                        queue1.merge(queue2, reverse)
                    }
                ))?
                .unwrap() as BoundedPriorityQueue<Self::Item>;

            Ok(queue.into())
        }
    }
```

Reply:
Sure thanks, I will try both tomorrow. 
Btw earlier I found a 3-party crate which provides an enhanced BinaryHeap, but I think we should stick to std library unless absolute necessary.
https://docs.rs/binary-heap-plus/0.2.0/binary_heap_plus/

Reply:
Summarizing, one option is more verbose but a little bit more efficient, the other is cleaner but a little less efficient/depending on what the compiler decides to do, requires extra compute steep, etc.

I have no strong opinion but since this will be done once and the verbosity here doesn't add complexity or makes code less clean I would implement the more efficient version in this case. but no strong opinion.

Reply:
I think there is a slight misconception here.  In Rust, 
```
collection.iter.map(|x| Reverse(x)).map(step1).map(step2) 
```
basically boils down to 
```
for x in collection {
    let x = Reverse(x);
    step1;
    step2
}
```

not like this:
```
for x in collection {
 let x = Reverse(x)
}
for x in collection {
step1
}
for x in collection {
step2
}
```
This is one of the amazing things in Rust.

Now let's see how iter way gets executed:
```
let heap = heap::new()
for x in collection {
    let x = Reverse(x); // It is not an allocation. Just a wrapper around T
    heap.append(x); // executes ord functions associated with Reverse instead of original Ord
}
```

Now how does enum way gets executed: 

```
let heap = heap::new();
for x in collection {
     heap.append(x); // all x have same functions but within each function function call, there is jump table lookup for enum match statements
}
```

Performance-wise, both should be near equal unless optimizer messes up jump table optimization with runtime checks in enum case which most likely won't happen. 

Now one disadvantage with the enum method is that now you have different API for take_ordered when compared to Spark. We should try to maintain our APIs as close to Spark as possible and make changes only when things are not possible here.

Reply:
But here:
`self.map(Fn!(|x| Reverse(x)))`

You are creating a new Rdd  (which is a new struct) that will add a new dependency to the DAG etc. which requires running the compute steep, the tasks etc etc etc. Is not like we are making an efficient single-pass iterator (not yet at least) that can be unrolled.

Reply:
I agree with this:

> Now one disadvantage with the enum method is that now you have different API for take_ordered when compared to Spark.

to a point...

If it's useful and clear I don't see what is the problem, for me it looks like that having the option of taking elements already sorted in a way I want by just passing one parameter and not having to run an extra steep is useful.

Also, this can be easily changed by having two different options if we wish (like with do with _num_partitions) etc.:

Reply:
The `Reverse(x)` is not the only difference between the 2 queues,  in the `maybe_replace_lowest(), each of them use `gt()` and `lt()` respectively, does the Reverse(x) also reverse the original results of the gt() and lt()?

Reply:
The only thing it does it's wrap the original type with a Reverse, afterwards when an operation that requires comparing or ordering that would be taken into account on every single method. So now that you say it, that would probably break maybe_replace in the way it's done now.

Reply:
@rajasekarv take in mind that in the Scala version the Ordering is imported implicitly:
```
def takeOrdered(num: Int)(implicit ord: Ordering[T])
```
So if you have in scope an implementation of Ordering for T that reverses the result this would be taken into account.

Reply:
`
You are creating a new rdd that will add a new dependency to the DAG etc. which requires running the compute steep, serializing, deserializing the tasks etc etc etc.
`
Actually, Rdd is just a wrapper around Iterators. If you look at the execution graph, it just takes one iterator and does something with it and creates another iterator. There is literally nothing happening in between. Basically bunch of compute() function calls during actual execution. In narrow dependency, there is no serialization. It is as same as Rust iterators we use normally. Only when there are shuffles in between serialization happens. 

This is why our way is very exciting than Spark's. For example, take the parquet example. I found no difference between solution based on Rust Iterators and based on RDDs. Both ran in exactly 15 secs and processed around 8.5GB worth of parquet data.

Reply:
Couldn't in the future potentially a task be executed in other executor even for narrow dependencies so it's not completely guaranteed (it should always try to use the preferred location but could happen is sent across workers) or this will always stay like this? If not I stand corrected.

In any case because what ugoa is saying it looks like it would be a breaking change anyway and cannot be approached that way. In one way or an other you are gonna require knowing what kind merging you are doing so that is gonna require a flag, enum or some sort of encoding at the type level (which would be even more complicated).

Reply:
@ugoa I don't think is correct we really need a `lt` alternative there in the `maybe_replace` method there if you actually use `Reverse` from the front up... it should be taken into account and then you wouldn't need to keep track if the item has been reversed or not, then what raja says would work fine. The reverse condition will be taken into account thereafter wrapping the type (eg. Reverse(4) < Reverse(3)).